src/multiint.cpp

```diff
@@ -2458,6 +2485,116 @@ static void loadMapSettings2()
 	}
 }
 
+static void randomizeLimit(const char *name)
+{
+	int stat = getStructStatFromName(name);
+	if (rand() % 2 == 0)
+	{
+		asStructureStats[stat].upgrade[0].limit = asStructureStats[stat].base.limit;
+	}
+	else
+	{
+		asStructureStats[stat].upgrade[0].limit = 0;
+	}
+}
+
+/* Generate random options */
+static void randomizeOptions()
+{
+	if (ingame.bHostSetup)
+	{
+		bLimiterLoaded = true;
+		if (!locked.scavengers)
+		{
+			game.scavengers = rand() % 2;
+			((MultichoiceWidget *)widgGetFromID(psWScreen, MULTIOP_GAMETYPE))->choose(game.scavengers);
+		}
+		if (!locked.alliances)
+		{
+			game.alliance = rand() % 4;
+			((MultichoiceWidget *)widgGetFromID(psWScreen, MULTIOP_ALLIANCES))->choose(game.alliance);
+		}
+		if (!locked.power)
+		{
+			game.power = rand() % 3;
+			((MultichoiceWidget *)widgGetFromID(psWScreen, MULTIOP_POWER))->choose(game.power);
+		}
+		if (!locked.bases)
+		{
+			game.base = rand() % 3;
+			((MultichoiceWidget *)widgGetFromID(psWScreen, MULTIOP_BASETYPE))->choose(game.base);
+		}
+
+		// Structure limits are simply 0 or max, only for NO_TANK, NO_CYBORG, NO_VTOL, NO_UPLINK, NO_LASSAT.
+		if (!bLimiterLoaded || !asStructureStats)
+		{
+			initLoadingScreen(true);
+			if (resLoad("wrf/limiter_data.wrf", 503))
+			{
+				bLimiterLoaded = true;
+			}
+			closeLoadingScreen();
+		}
+		for (int i = 0; i < ARRAY_SIZE(limitIcons) - 1; ++i)	// skip last item, MPFLAGS_FORCELIMITS
+		{
+			randomizeLimit(limitIcons[i].stat);
+		}
+		if (rand() % 2 == 0)
+		{
+			ingame.flags |= MPFLAGS_FORCELIMITS;
+		}
+		else
+		{
+			ingame.flags &= ~MPFLAGS_FORCELIMITS;
+		}
+		createLimitSet();
+		applyLimitSet();
+		updateLimitFlags();
+		updateLimitIcons();
+
+		// Pick a map for a number of players and tech level
+		game.techLevel = current_tech = (rand() % 4) + 1;
+		LEVEL_LIST levels;
+		do
+		{
+			// don't kick out already joined players because of randomize
+			int players = countJoiningHumans();
+			int minimumPlayers = std::max(players, 2);
+			current_numplayers = minimumPlayers;
+			if (minimumPlayers < MAX_PLAYERS_IN_GUI)
+			{
+				current_numplayers += (rand() % (MAX_PLAYERS_IN_GUI - minimumPlayers)) + minimumPlayers;
```

Comment:
@KJeff01 Here I set the range of random players, it should be number of humans as minumum, plus additional AIs up to MAX. Have you seen any problems?

Reply:
`MAX_PLAYERS_IN_GUI` is always 10, I think. `game.maxPlayers` has the true maximum amount of players supported by a map. AI count shouldn't matter if switching maps.

See, with my test, I was one a 8 player map, picked the 5th slot with my other Warzone instance and then randomized again and it landed on a 4-player map where the non-hosting human slot disappeared and it was just the host and 3 AI. The other player turned into a ghost that could show up again after more map changes.

Reply:
Good catch. This was tricky to fix because of scavenger slots and all. I ended up implementing random start positions instead of trying to swap and rearrange the players.